src/resolve/test-dns-server.c

```diff
@@ -0,0 +1,820 @@
+}
+
+TEST(dns_server_possible_feature_level_udp_failed) {
+        _cleanup_(server_env_teardown) ServerEnv env;
+        server_env_setup(&env, false);
+
+        env.server->n_failed_udp = 5;
+        env.server->possible_feature_level = DNS_SERVER_FEATURE_LEVEL_EDNS0;
+
+        dns_server_possible_feature_level(env.server);
+
+        ASSERT_EQ(env.server->possible_feature_level, DNS_SERVER_FEATURE_LEVEL_UDP);
+}
+
+TEST(dns_server_possible_feature_level_tcp_failed_truncated) {
+        _cleanup_(server_env_teardown) ServerEnv env;
+        server_env_setup(&env, false);
+
+        env.server->n_failed_tcp = 5;
+        env.server->packet_truncated = true;
+        env.server->possible_feature_level = DNS_SERVER_FEATURE_LEVEL_EDNS0;
+
+        dns_server_possible_feature_level(env.server);
+
+        ASSERT_EQ(env.server->possible_feature_level, DNS_SERVER_FEATURE_LEVEL_UDP);
+}
+
+/* ================================================================
+ * dns_server_string(), dns_server_string_full()
+ * ================================================================ */
+
+TEST(dns_server_string) {
+        _cleanup_(server_env_teardown) ServerEnv env;
+        server_env_setup(&env, false);
+
+        ASSERT_STREQ(dns_server_string(env.server), "192.168.1.128");
+}
+
+TEST(dns_server_string_full) {
+        _cleanup_(server_env_teardown) ServerEnv env;
+        server_env_setup(&env, false);
+
+        const char *str = dns_server_string_full(env.server);
+        ASSERT_STREQ(str, "192.168.1.128:53#server.local");
+}
+
+/* ================================================================
+ * dns_server_dnssec_supported()
+ * ================================================================ */
+
+TEST(dns_server_dnssec_supported) {
+        _cleanup_(server_env_teardown) ServerEnv env;
+        server_env_setup(&env, false);
+
+        env.manager.dnssec_mode = DNSSEC_YES;
+        ASSERT_TRUE(dns_server_dnssec_supported(env.server));
+
+        env.manager.dnssec_mode = DNSSEC_NO;
+        env.server->packet_bad_opt = true;
+        ASSERT_FALSE(dns_server_dnssec_supported(env.server));
+
+        env.server->packet_bad_opt = false;
+        env.server->packet_rrsig_missing = true;
+        ASSERT_FALSE(dns_server_dnssec_supported(env.server));
+
+        env.server->packet_rrsig_missing = false;
+        env.server->packet_do_off = true;
+        ASSERT_FALSE(dns_server_dnssec_supported(env.server));
+
+        env.server->packet_do_off = false;
+        env.server->n_failed_tcp = 5;
+        ASSERT_FALSE(dns_server_dnssec_supported(env.server));
+
+        env.server->n_failed_tcp = 0;
+        ASSERT_TRUE(dns_server_dnssec_supported(env.server));
+}
+
+/* ================================================================
+ * dns_server_dump()
+ * ================================================================ */
+
+static void check_dump_contents(FILE *f, const char **expected, size_t n) {
+        char *actual[n];
```

Comment:
this might trigger a static analyzer like coverity - assert that n > 0 should be enough to avoid that